segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,352 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+                                    txn.delete(segmentMetadata.getName());
+                                    return txn.commit();
+                                }, storageExecutor)
+                                .handleAsync((v, e) -> {
+                                    if (null != e) {
+                                        log.error(String.format("%s deleteGarbage - Could not delete metadata for garbage segment=%s.",
+                                                traceObjectId, streamSegmentName), e);
+                                        return true;
+                                    }
+                                    return false;
+                                }, storageExecutor)
+                                .thenComposeAsync(failed -> {
+                                    if (failed) {
+                                        if (taskInfo.getAttempts() < config.getGarbageCollectionMaxAttempts()) {
+                                            val attempts = taskInfo.attempts + 1;
+                                            SLTS_GC_SEGMENT_RETRY.inc();
+                                            return addSegmentToGarbage(taskInfo.toBuilder().attempts(attempts).build());
+                                        } else {
+                                            SLTS_GC_SEGMENT_FAILED.inc();
+                                            log.info("{}: deleteGarbage - could not delete after max attempts segment={}.", traceObjectId, taskInfo.getName());
+                                            return failTask(taskInfo);
+                                        }
+                                    } else {
+                                        SLTS_GC_SEGMENT_PROCESSED.inc();
+                                        return CompletableFuture.completedFuture(null);
+                                    }
+                                }, storageExecutor);
+                    }
+                }, storageExecutor);
     }
 
     /**
-     * Adds given chunk to list of garbage chunks.
-     *
-     * @param chunkToDelete Name of the chunk to delete.
-     * @param startTime Start time.
-     * @param attempts Number of attempts to delete this chunk so far.
+     * Delete block index entries for given chunk.
      */
-    void addToGarbage(String chunkToDelete, long startTime, int attempts) {
-        if (queueSize.get() < config.getGarbageCollectionMaxQueueSize()) {
-            garbageChunks.add(new GarbageChunkInfo(chunkToDelete, startTime, attempts));
-            queueSize.incrementAndGet();
-        } else {
-            log.debug("{}: deleteGarbage - Queue full. Could not delete garbage. chunk {}.", traceObjectId, chunkToDelete);
+    void deleteBlockIndexEntriesForChunk(MetadataTransaction txn, String segmentName, long startOffset, long endOffset) {
+        val firstBlock = startOffset / config.getIndexBlockSize();
```

Comment:
Would it be advisable to add some preconditions on this calculation (e.g., `firstBlock>0`, `config.getIndexBlockSize()>0`)?

Reply:
I think its not needed here - these are checked independently in other places.
Eg. https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java#L212